### PR TITLE
Overhaul training & optimisation w/ new methods & classes

### PR DESCRIPTION
**Pull request by SBouchard01**
This PR reworks the `train` & `optimise` modules to add flexibility in the call, with options to save the best models.

The `train.fit` method is kept as a deprecated one for compatibility.

`optimise` now has a `train_best_model` method, that is immediately deprecated, as some randomness happens inside - I recommend extracting the best checkpoint from the study instead (the `StudyFCN` method has a parameter to do so easily).

### CODE REVIEW

File: sunbird/emulators/train.py [hunk 1]
```diff
@@ -1,10 +1,341 @@
...
+        
+        # Retrieve the best validation loss from the EarlyStopping callback
+        best_val_loss = None
+        for cb in self.callbacks:
+            if isinstance(cb, EarlyStopping):
+                best_val_loss = cb.best_score.item()
+                break
+        if best_val_loss is None:
+            best_val_loss = self.callback_metrics.get('val_loss', None)
+
```
[the remark this answers is not in the document]
Comment (SBouchard01) [reply sunbird/emulators/train.py]: If best_val_loss is still None after line 100, it means that no callback was created with the 'val_loss' metric, and this is the expected behavior.
Adding a warning just in case though.

File: sunbird/emulators/train.py [hunk 2]
```diff
@@ -1,10 +1,341 @@
...
+        super().fit(*args, **kwargs)
+        
+        # Retrieve the best validation loss from the EarlyStopping callback
+        best_val_loss = None
+        for cb in self.callbacks:
+            if isinstance(cb, EarlyStopping):
+                best_val_loss = cb.best_score.item()
+                break
+        if best_val_loss is None:
+            best_val_loss = self.callback_metrics.get('val_loss', None)
```
Comment (SBouchard01) [inline sunbird/emulators/train.py]: There should be a cleaner way to ensure the best validation loss is the output of `fit()`, but I don't know yet how to implement it (except if we enforce the callbacks at init...)
